rfcs/2021-09-01-8547-accept-metrics-in-datadog-agent-source.md

```diff
@@ -0,0 +1,277 @@
+dogstatsd and from go/python code. It mostly aims at reducing the amount of metrics samples sent by the Agent.
+
+[Veneur](https://github.com/stripe/veneur#global-aggregation) offers an aggregation feature, but it does not support
+sketches/distribution per se. It requires what is called a central veneur, that would compute aggregated value for
+selected metrics and some percentile. Some aspects of this solution could be seen as an
+[alternative](#flattening-sketches) approach. However this approach has two major drawbacks: it relies on a central
+service for aggregation and it does not support sketches.
+
+## Alternatives
+
+### For transport between Agents and Vector
+The use an alternate protocol between Datadog Agents and Vector (Like Prometheus, Statds, OpenTelemetry or Vector own
+protocol) could be envisionned. This would call for a significant, yet possible with the current Agent architecture,
+addition, those changes would mostly be located in the
+[forwarder](https://github.com/DataDog/datadog-agent/tree/main/pkg/forwarder) and
+[serializer](https://github.com/DataDog/datadog-agent/tree/main/pkg/serializer) logic. This would imply a hugh chunk of
+work on the Agent side, require update to use the feature, probably also require some work on Vector side. This is not
+something that aligns well with the purpose of the Datadog Agent. This would also add a risk of losing information
+because of protocol conversion.
+
+### Flattening sketches
+For sketches, we could flatten sketches and compute usual derived metrics (min/max/average/count/some percentiles) and
+send those as gauge/count, but it would prevent (or at least impact) existing distribution/sketches users. Moreover if
+instead of sketches only derived metrics are used a lot of the tagging flexiblity will be lost. By submitting tagged
+sketches to the Datadog intake, any tag selector can be used to compute a distribution based on the sketches that bear
+matching tag. This cannot be done without sending sketches. But flattening sketches would have the benefit of simplify
+the implementation in Vector and remove the prerequisite of having sketches support inside Vector.
+
+### For Request routing
+This is point is still being discussed. As stated in [Outstanding Questions](#outstanding-questions) there are 3
+alternatives solution currently envisionned. The current suggested one is to left the Agent untouched and handle
+everything in Vector. Alternatives to that are:
+1. pushing for a Datadog Agent change with a new override (let's say `metrics_dd_url` that would only divert request to
+   `/api/v1/series` & `/api/beta/sketches` to a specific endpoints. This however highlights that Vector & the Datadog
+   Agent will then need to follow a compatbility matrix:
+  * If the Agent is upgraded and introduce a new metric route (let's say for v2 intake migration), the user will need to
+    also upgrade Vector so it supports this route.
```

Comment:
Won't this also be the case even with the current suggestion? If the Datadog agent starts sending on a new metrics route, we'd need to add support to Vector to be able to process it rather than just proxy it.

Reply:
Indeed, but my point was if a new metric route is introduced in the Agent:
* If the url routing is happening in haproxy, then the default routing is likely to get request to this route sent directly to datadog, preventing metric loss
* if the url routing is happening in vector, the payload will also be sent to datadog following the "unknown/unsupported payload" logic, preventing metric loss
* if the url routing is happening in the Agent, then a non-upgraded-vector will receive payload it has no idea what to do with and causes metric loss _unless we decide to implement a fallback path for unknown payload_ 

I'm not exactly sure how often this would happen (not very I guess) but there might a minor risk of loosing data if we only support a strict list of route in Vector without having a fallback forwarding logic for unknown payload (having it might raise some security concern however).